lnwallet/chancloser/rbf_coop_transitions.go

```diff
@@ -0,0 +1,1109 @@
+	// has confirmed, so we can move to our terminal state early.
+	case *SpendEvent:
+		return &CloseStateTransition{
+			NextState: &CloseFin{
+				transitionEvent: msg,
+				ConfirmedTx:     msg.Tx,
+			},
+		}, nil
+
+	// If we get an OfferReceived event, then the channel is flushed from
+	// the PoV of the remote party. However, due to propagation delay or
+	// concurrency, we may not have received the ChannelFlushed event yet.
+	// In this case, we'll stash the event and wait for the ChannelFlushed
+	// event.
+	case *OfferReceivedEvent:
+		chancloserLog.Infof("ChannelPoint(%v): received remote offer "+
+			"early, stashing...", env.ChanPoint)
+
+		c.EarlyRemoteOffer = fn.Some(*msg)
+
+		// TODO(roasbeef): unit test!
+
+		// We'll perform a noop update so we can wait for the actual
+		// channel flushed event.
+		return &CloseStateTransition{
+			NextState: c,
+		}, nil
+
+	// If we receive the ChannelFlushed event, then the coast is clear so
+	// we'll now morph into the dual peer state so we can handle any
+	// messages needed to drive forward the close process.
+	case *ChannelFlushed:
+		// Both the local and remote losing negotiation needs the terms
+		// we'll be using to close the channel, so we'll create them
+		// here.
+		closeTerms := CloseChannelTerms{
+			ShutdownScripts:  c.ShutdownScripts,
+			ShutdownBalances: msg.ShutdownBalances,
+		}
+
+		chancloserLog.Infof("ChannelPoint(%v): channel flushed! "+
+			"proceeding with co-op close", env.ChanPoint)
+
+		// Now that the channel has been flushed, we'll mark on disk
+		// that we're approaching the point of no return where we'll
+		// send a new signature to the remote party.
+		//
+		// TODO(roasbeef): doesn't actually matter if initiator here?
+		if msg.FreshFlush {
+			err := env.ChanObserver.MarkCoopBroadcasted(nil, true)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// If an ideal fee rate was specified, then we'll use that,
+		// otherwise we'll fall back to the default value given in the
+		// env.
+		idealFeeRate := c.IdealFeeRate.UnwrapOr(env.DefaultFeeRate)
+
+		// We'll then use that fee rate to determine the absolute fee
+		// we'd propose.
+		//
+		// TODO(roasbeef): need to sign the 3 diff versions of this?
+		localTxOut, remoteTxOut := closeTerms.DeriveCloseTxOuts()
+		absoluteFee := env.FeeEstimator.EstimateFee(
+			env.ChanType, localTxOut, remoteTxOut,
+			idealFeeRate.FeePerKWeight(),
+		)
+
+		chancloserLog.Infof("ChannelPoint(%v): using ideal_fee=%v, "+
+			"absolute_fee=%v", env.ChanPoint, idealFeeRate,
+			absoluteFee)
+
+		var (
+			internalEvents []ProtocolEvent
+			newEvents      fn.Option[protofsm.EmittedEvent[ProtocolEvent]]
+		)
+
+		// If we received a remote offer early from the remote party,
+		// then we'll add that to the set of internal events to emit.
+		c.EarlyRemoteOffer.WhenSome(func(offer OfferReceivedEvent) {
+			internalEvents = append(internalEvents, &offer)
+		})
```

Comment:
why not just buffer the events in the driver rather than ingesting it into the state machine prematurely and having to reinterpret it as an internal event? I generally think that the internal event architecture is a bit of a rough edge of the design and I think this makes tracing how the event gets processed even more convoluted since it originated as an external event but is being replayed as an internal one.

Reply:
How would you suggest we buffer it in the driver? 

As discussed before, if we don't want to emit internal events, then we end up making each state transition larger, as it covers more of the state space. By emitting the internal events, I was able to keep each state transition as small as possible. 

In this particular instance, it's a way to handle early events, without the driver needing to be burdened with handling them. This can happen if the other peer sends something thing (we handle it to be liberal in what we accept), or due to other asynchronous actions elsewhere in the daemon.

Reply:
I misunderstood what this was accomplishing, feel free to ignore. I thought this was the result of parallel execution and not serial processing where we want to more resiliently handle an out-of-order offer from the peer.